### ISSUE TITLE
Consider expanding JIT Asm support

### ISSUE THREAD

**Issue by tannergooding** [ashmind/SharpLab#384]
Currently the `JIT Asm` results target only targets Desktop CLR and does not let you select 32-bit vs 64-bit (you can select 32-bit vs 64-bit for the main C# compiler only).

Given that most runtime improvements are happening on CoreCLR, it would be beneficial if you could select the target runtime for the `JIT Asm` target (if nothing else).

Given that the compiler can target any cpu target (`AnyCPU`, `x86`, `x64`, `ARM64`, etc), it would be also be beneficial if you could select which assembly output you want.

This effectively gives you the following options:
* Language (C# vs VB vs F#)
  * Language Version would also be nice (i.e. LangVersion=7.3 vs Preview vs Latest)
* Compiler Version (Latest Release vs Preview Branches)
* Results (all of this applies to JIT Asm, I think there are sandboxing concerns for things like Run)
* Runtime (Desktop vs CoreCLR vs Maybe Mono)
  * Runtime Version would also be nice
  * Target CPU would be nice (x86 vs x64 vs ARM vs ARM64)
* Configuration (Debug vs Release)

**Comment by ashmind** [comment]
Thanks for the feedback!

I do want to support CoreCLR, however I need `AssemblyLoadContext.Unload` which is only available in .NET Core 3.0. I might look at using the preview version in the meantime, but I can't promise anything specific earlier than 3.0.

> Language Version would also be nice (i.e. LangVersion=7.3 vs Preview vs Latest)

What's the use case for this? I considered it before, but couldn't find a compelling reason.

> Compiler Version (Latest Release vs Preview Branches)

I don't think I can provide Core vs x86 vs x64 for each branch as currently each branch is separate Azure WebApp. Providing those options will turn each branch into 3 WebApps (not really a blocker, but I feel it will be too hard to manage).

> Target CPU would be nice (x86 vs x64 vs ARM vs ARM64)

It's interesting, but I don't think Azure provides ARM/ARM64 hosting at the moment.
I wonder if there is an emulator though, I'll investigate.

**Comment by tannergooding** [comment]
> I might look at using the preview version in the meantime, but I can't promise anything specific earlier than 3.0.

That sounds like a reasonable trade-off; although, I imagine you could also achieve this via reflection only load, System.Reflection.Metadata, or some other tooling on prior versions. That is, for the common case of just getting the IL or JIT disassembly, I don't believe you need `AssemblyLoadContext` to achieve most of the functionality. You should only need that for running the code which isn't something I'm asking for here.

> What's the use case for this? I considered it before, but couldn't find a compelling reason.

One of the most useful features of sharplab, in my mind, is being able to easily see the compiled IL and JIT disassembly SxS with the C# code. If there was a VS extension to do this, I would definitely have it installed.

Being able to select the language version allows you to explicitly see minor differences in codegen (both IL and assembly) that different language features or versions bring in.

> I don't think I can provide Core vs x86 vs x64 for each branch as currently each branch is separate Azure WebApp.

Why would this need to be a separate web app rather than just a configuration switch? The compiler supports generating IL for 'AnyCPU', 'X86', 'X64', etc... regardless of the platform you are running on. Likewise, there exists tooling for getting the disassembly for a given CPU target regardless of what CPU you are running against.
* Benchmark.NET has some functionality built-in
* dotnet/jitutils provides probably the best support for this; although it requires a checked build of CoreCLR to function; and sometimes setting the AltJIT to get codegen for other CPUs...
* There are likely others as well, but I'm not familiar with them

Basically, I don't think you really need to have a host CPU that is the same as the target platform outside the 'Run' scenario.

**Comment by ashmind** [comment]
> Being able to select the language version allows you to explicitly see minor differences in codegen (both IL and assembly) that different language features or versions bring in.

But is there actually a difference within a certain compiler release? From what I remember in Roslyn code, `LanguageVersion` makes compiler reject newer syntax, but I don't know of any case where IL will be generated differently for the same feature.

> I don't believe you need `AssemblyLoadContext` to achieve most of the functionality.  You should only need that for running the code which isn't something I'm asking for here.
> Basically, I don't think you really need to have a host CPU that is the same as the target platform outside the 'Run' scenario. 

Formally you are right, but only reasonable JIT approach I found requires runnable code.
`dotnet/jitutils` wasn't easily usable last time I looked (I might need to recheck). As for Benchmark.NET -- can you point to the specific code/feature? I'll investigate.

**Comment by svick** [comment]
@ashmind 

> But is there actually a difference within a certain compiler release? From what I remember in Roslyn code, `LanguageVersion` makes compiler reject newer syntax, but I don't know of any case where IL will be generated differently for the same feature.

The one case I'm aware of is the `peverify-compat` mode, which is enabled by default with LangVersion < 7.2, see https://github.com/dotnet/roslyn/pull/22772.

> As for Benchmark.NET -- can you point to the specific code/feature? I'll investigate.

BDN has a [`DiassemblyDiagnoser`](https://benchmarkdotnet.org/articles/features/disassembler.html). As far as I can tell, on Windows, [it uses ClrMD to attach to a process and disassembles the code from that](https://github.com/dotnet/BenchmarkDotNet/blob/a2bddf3/src/BenchmarkDotNet.Disassembler.x64/Program.cs). So I think it also requires runnable code.

**Comment by ashmind** [comment]
> As far as I can tell, on Windows, it uses ClrMD to attach to a process and disassembles the code from that. So I think using the same approach would also require runnable code.

Hmm, no surprise, from dotnet/BenchmarkDotNet#437 it seems they are at least partially based on SharpLab implementation (_"As in TryRoslyn ..."_).

**Comment by tannergooding** [comment]
> BDN has a DiassemblyDiagnoser. As far as I can tell, on Windows, it uses ClrMD to attach to a process and disassembles the code from that. So I think using the same approach would also require runnable code.

Ah interesting. I had thought there was a different approach that let it work independently, but I must be misremembering.

> dotnet/jitutils wasn't easily usable last time I looked (I might need to recheck)

There have been a number of improvements here, but the primary thing is that it still requires a debug/checked build of the JIT dll (`corjit`). Maybe there is something that could be done to make that process easier so things don't have to be built locally for .NET Core (similar to how you are now pulling down pre-built packages for Roslyn)...

**Comment by am11** [comment]
.NET Core support has been included since https://github.com/ashmind/SharpLab/commit/e660366fe326885f8906f9a5a798771ef468573d.
The dropdown at https://sharplab.io/ lists the new runtimes:

<img width="30%" src="https://user-images.githubusercontent.com/3840695/59293410-d267c380-8c87-11e9-8e14-7e448e26a7ea.png" />

**Comment by MichalPetryka** [comment]
Having Mono runtime could be useful for checking how the codegen would look in game engines that use it, for example Unity. Would be nice if it was added.
